From 2e4735c8b63fed1e34d1dfb11c665544494320d8 Mon Sep 17 00:00:00 2001
From: Rahul Akolkar <rahul@apache.org>
Date: Mon, 6 Jun 2011 22:49:05 +0000
Subject: [PATCH] Port r1132819 from trunk. SCXML-161 Transition leaving a
 child state of parallel incorrect. Fix by accounting for non-composite
 regions. Test case provided by Enrico Nardelli <nardelli at mat dot uniroma2
 dot it>.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/scxml/branches/J6@1132823 13f79535-47bb-0310-9956-ffa450edef68
---
 .../org/apache/commons/scxml/SCXMLHelper.java |  2 +-
 .../commons/scxml/SCXMLExecutorTest.java      | 17 ++++++++-
 .../apache/commons/scxml/transitions-06.xml   | 38 +++++++++++++++++++
 3 files changed, 54 insertions(+), 3 deletions(-)
 create mode 100644 src/test/java/org/apache/commons/scxml/transitions-06.xml

diff --git a/src/main/java/org/apache/commons/scxml/SCXMLHelper.java b/src/main/java/org/apache/commons/scxml/SCXMLHelper.java
index 222f77211..cf52f20fe 100644
--- a/src/main/java/org/apache/commons/scxml/SCXMLHelper.java
+++ b/src/main/java/org/apache/commons/scxml/SCXMLHelper.java
@@ -250,7 +250,7 @@ public static Set<TransitionTarget> getStatesExited(final Transition t,
                 for (TransitionTarget tt : par.getChildren()) {
                     State s = (State) tt;
                     for (TransitionTarget a : currentStates) {
-                        if (isDescendant(a, s)) {
+                        if (isDescendant(a, s) || a == s) {
                             //a is affected
                             boolean added = false;
                             added = allStates.add(a);
diff --git a/src/test/java/org/apache/commons/scxml/SCXMLExecutorTest.java b/src/test/java/org/apache/commons/scxml/SCXMLExecutorTest.java
index 4e4e2bf78..89cac8b2e 100644
--- a/src/test/java/org/apache/commons/scxml/SCXMLExecutorTest.java
+++ b/src/test/java/org/apache/commons/scxml/SCXMLExecutorTest.java
@@ -41,7 +41,7 @@ public SCXMLExecutorTest(String name) {
     // Test data
     private URL microwave01jsp, microwave02jsp, microwave01jexl,
         microwave02jexl, microwave03jexl, microwave04jexl, microwave05jexl, transitions01,
-        transitions02, transitions03, transitions04, transitions05, prefix01, send01, send02;
+        transitions02, transitions03, transitions04, transitions05, transitions06, prefix01, send01, send02;
     private SCXMLExecutor exec;
 
     /**
@@ -73,6 +73,8 @@ public void setUp() {
             getResource("org/apache/commons/scxml/transitions-04.xml");
         transitions05 = this.getClass().getClassLoader().
             getResource("org/apache/commons/scxml/transitions-05.xml");
+        transitions06 = this.getClass().getClassLoader().
+            getResource("org/apache/commons/scxml/transitions-06.xml");
         prefix01 = this.getClass().getClassLoader().
             getResource("org/apache/commons/scxml/prefix-01.xml");
         send01 = this.getClass().getClassLoader().
@@ -88,7 +90,7 @@ public void setUp() {
     public void tearDown() {
         microwave01jsp = microwave02jsp = microwave01jexl = microwave02jexl =
             microwave04jexl = microwave05jexl = transitions01 = transitions02 = transitions03 =
-            transitions04 = transitions05 = prefix01 = send01 = send02 = null;
+            transitions04 = transitions05 = transitions06 = prefix01 = send01 = send02 = null;
     }
 
     /**
@@ -235,6 +237,17 @@ public void testSCXMLExecutorTransitions05Sample() throws Exception {
         SCXMLTestHelper.assertPostTriggerState(exec, "foo", "end");
     }
 
+    public void testSCXMLExecutorTransitions06Sample() throws Exception {
+        SCXML scxml = SCXMLTestHelper.parse(transitions06);
+        assertNotNull(scxml);
+        exec = SCXMLTestHelper.getExecutor(scxml);
+        assertNotNull(exec);
+        SCXMLTestHelper.assertPostTriggerStates(exec, "start", new String[]{"one", "two"});
+        SCXMLTestHelper.assertPostTriggerState(exec, "onetwo_three", "three");
+        SCXMLTestHelper.assertPostTriggerStates(exec, "three_one", new String[]{"one", "two"});
+        SCXMLTestHelper.assertPostTriggerState(exec, "two_four", "four");
+    }
+
     public void testSend01Sample() throws Exception {
         exec = SCXMLTestHelper.getExecutor(send01);
         assertNotNull(exec);
diff --git a/src/test/java/org/apache/commons/scxml/transitions-06.xml b/src/test/java/org/apache/commons/scxml/transitions-06.xml
new file mode 100644
index 000000000..7400359d8
--- /dev/null
+++ b/src/test/java/org/apache/commons/scxml/transitions-06.xml
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+-->
+<scxml xmlns="http://www.w3.org/2005/07/scxml" version="1.0" initial="init">
+       <state id="init">
+               <transition event="start" target="onetwo" />
+       </state>
+       <parallel id="onetwo">
+               <transition event="onetwo_three" target="three" />
+               <state id="one">
+               </state>
+               <state id="two">
+                       <transition event="two_four" target="four" />
+               </state>
+       </parallel>
+       <state id="three">
+               <transition event="three_one" target="one" />
+               <transition event="three_four" target="four" />
+       </state>
+       <state id="four">
+               <transition event="four_onetwo" target="onetwo" />
+               <transition event="four_three" target="three" />
+       </state>
+</scxml>
